File: data/derived_cot/rq1_traces/translation/r1/java_to_cpp/code_output/ExpressionCalculator.cpp
#include <deque>
#include <string>
#include <cctype>
#include <algorithm>
#include <cmath>
#include <stdexcept>

class ExpressionCalculator {
private:
    std::deque<std::string> postfixStack;

    bool isOperator(char ch) {
        return ch == '+' || ch == '-' || ch == '*' || ch == '/' || ch == '%';
    }

    bool compare(char op1, char op2) {
        char cur_op = op1 == '%' ? '/' : op1;
        char peek_op = op2 == '%' ? '/' : op2;
        int operat_priority[8] = {0, 3, 2, 1, -1, 1, 0, 2};
        int idx_peek = peek_op - 40;
        int idx_cur = cur_op - 40;
        if (idx_peek < 0 || idx_peek >= 8 || idx_cur < 0 || idx_cur >= 8) {
            return false;
        }
        return operat_priority[idx_peek] >= operat_priority[idx_cur];
    }

    double _calculate(double a, double b, char op) {
        switch (op) {
            case '+': return a + b;
            case '-': return a - b;
            case '*': return a * b;
            case '/': return a / b;
            case '%': return std::fmod(a, b);
            default:
                throw std::invalid_argument("Unsupported operator: " + std::string(1, op));
        }
    }

    std::string transform(std::string expression) {
        expression.erase(std::remove(expression.begin(), expression.end(), ' '), expression.end());
        for (char& c : expression) {
            if (c == '-') {
                c = '~';
            }
        }
        if (!expression.empty() && expression[0] == '~' && expression.size() > 1 && expression[1] == '(') {
            expression[0] = '-';
            return "0" + expression;
        }
        return expression;
    }

    void prepare(const std::string& expression) {
        postfixStack.clear();
        std::deque<char> operatorStack;
        int length = expression.length();
        for (int i = 0; i < length; i++) {
            char ch = expression[i];
            if (std::isdigit(static_cast<unsigned char>(ch)) || ch == '.') {
                std::string num;
                while (i < length && (std::isdigit(static_cast<unsigned char>(expression[i])) || expression[i] == '.')) {
                    num += expression[i];
                    i++;
                }
                i--;
                postfixStack.push_back(num);
            } else if (ch == '(') {
                operatorStack.push_front(ch);
            } else if (ch == ')') {
                while (!operatorStack.empty() && operatorStack.front() != '(') {
                    postfixStack.push_back(std::string(1, operatorStack.front()));
                    operatorStack.pop_front();
                }
                if (!operatorStack.empty()) {
                    operatorStack.pop_front();
                }
            } else if (isOperator(ch)) {
                while (!operatorStack.empty() && isOperator(operatorStack.front()) && !compare(operatorStack.front(), ch)) {
                    postfixStack.push_back(std::string(1, operatorStack.front()));
                    operatorStack.pop_front();
                }
                operatorStack.push_front(ch);
            }
        }
        while (!operatorStack.empty()) {
            postfixStack.push_back(std::string(1, operatorStack.front()));
            operatorStack.pop_front();
        }
    }

    double evaluatePostfix() {
        std::deque<double> stack;
        for (const std::string& token : postfixStack) {
            if (token.length() == 1 && isOperator(token[0])) {
                if (stack.size() < 2) {
                    throw std::runtime_error("Insufficient operands for operator");
                }
                double b = stack.front();
                stack.pop_front();
                double a = stack.front();
                stack.pop_front();
                stack.push_front(_calculate(a, b, token[0]));
            } else {
                try {
                    size_t pos;
                    double num = std::stod(token, &pos);
                    if (pos != token.length()) {
                        throw std::invalid_argument("Invalid token: " + token);
                    }
                    stack.push_front(num);
                } catch (const std::invalid_argument& e) {
                    throw std::invalid_argument("Invalid token: " + token);
                }
            }
        }
        if (stack.size() != 1) {
            throw std::runtime_error("Malformed expression");
        }
        return stack.front();
    }

public:
    ExpressionCalculator() {}

    double calculate(const std::string& expression) {
        std::string transformedExpression = transform(expression);
        prepare(transformedExpression);
        return evaluatePostfix();
    }
};

File: data/derived_cot/rq1_traces/translation/r1/java_to_py/code_output/ExpressionCalculator.py
import collections

class ExpressionCalculator:
    def __init__(self):
        self.postfix_stack = collections.deque()
    
    def calculate(self, expression):
        transformed = self.transform(expression)
        self.prepare(transformed)
        return self.evaluate_postfix()
    
    def transform(self, expression):
        expr = expression.replace(" ", "")
        expr = expr.replace("-", "~")
        if expr.startswith("~(") and len(expr) > 2:
            return "0-" + expr[1:]
        return expr
    
    def prepare(self, expression):
        self.postfix_stack.clear()
        operator_stack = collections.deque()
        n = len(expression)
        i = 0
        while i < n:
            ch = expression[i]
            if ch.isdigit() or ch == '.':
                num = []
                j = i
                while j < n and (expression[j].isdigit() or expression[j] == '.'):
                    num.append(expression[j])
                    j += 1
                self.postfix_stack.append(''.join(num))
                i = j - 1
            elif ch == '(':
                operator_stack.append(ch)
            elif ch == ')':
                while operator_stack and operator_stack[-1] != '(':
                    self.postfix_stack.append(operator_stack.pop())
                if operator_stack:
                    operator_stack.pop()
            elif self.is_operator(ch):
                while operator_stack and operator_stack[-1] != '(' and self.is_operator(operator_stack[-1]) and not self.compare(operator_stack[-1], ch):
                    self.postfix_stack.append(operator_stack.pop())
                operator_stack.append(ch)
            i += 1
        
        while operator_stack:
            self.postfix_stack.append(operator_stack.pop())
    
    def is_operator(self, ch):
        return ch in ['+', '-', '*', '/', '%']
    
    def compare(self, op1, op2):
        op1_mapped = '/' if op1 == '%' else op1
        op2_mapped = '/' if op2 == '%' else op2
        priority = [0, 3, 2, 1, -1, 1, 0, 2]
        idx1 = ord(op1_mapped) - 40
        idx2 = ord(op2_mapped) - 40
        if idx1 < 0 or idx1 >= len(priority) or idx2 < 0 or idx2 >= len(priority):
            return False
        return priority[idx2] >= priority[idx1]
    
    def evaluate_postfix(self):
        stack = collections.deque()
        for token in self.postfix_stack:
            if len(token) == 1 and self.is_operator(token[0]):
                b = stack.pop()
                a = stack.pop()
                stack.append(self._calculate(a, b, token[0]))
            else:
                stack.append(float(token))
        return stack.pop()
    
    def _calculate(self, a, b, operator):
        if operator == '+':
            return a + b
        elif operator == '-':
            return a - b
        elif operator == '*':
            return a * b
        elif operator == '/':
            return a / b
        elif operator == '%':
            return a % b
        else:
            raise ValueError(f"Unsupported operator: {operator}")